keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,469 @@
+
+If the user set both `percent` and `pods` to `0`, it means that the corresponding resource (e.g. `deploy`, `rs`) should not be scaled in that direction.
+
+[The motivation to pick the largest constraint]: #the-motivation-to-pick-the-largest-constraint-concept
+[Default Values]: #default-values
+[Stabilization Window]: #stabilization-window
+
+### User Stories
+
+#### Story 1: Scale Up As Fast As Possible
+
+This mode is essential when you want to respond to a traffic increase quickly.
+
+Create an HPA with the following configuration:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900`    (i.e., to increase the number of pods 10 times per minute is ok).
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up with the following number of pods:
+
+    1 -> 10 -> 100 -> 1000
+
+So, it can reach `maxReplicas` very fast.
+
+Scale down will be done a usual way (check stabilization window in the [Stabilization Window][] section below and the [Algorithm details][] in the official HPA documentation)
+
+[Stabilization Window]: #stabilization-window
+[Algorithm details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 2: Scale Up As Fast As Possible, Scale Down Very Gradually
+
+This mode is essential when you don’t want to risk scaling down very quickly.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900` (i.e. increase number of pods 10 times per minute is ok).
+  - `scaleDown`:
+    - `pods = 1`
+    - `periodSeconds = 600` (i.e., scale down one pod every 10 min)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up with the following number of pods:
+
+    1 -> 10 -> 100 -> 1000
+
+So, it can reach `maxReplicas` very fast.
+
+Scaling down will be by one pod each 10 min:
+
+    1000 -> 1000 -> 1000 -> … (7 more min) -> 999
+
+#### Story 3: Scale Up Very Gradually, Usual Scale Down Process
+
+This mode is essential when you want to increase capacity, but you want it to be very pessimistic.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `pods = 1`    (i.e., increase only by one pod per minute)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up very gradually:
+
+    1 -> 2 -> 3 -> 4
+
+Scale down will be done a usual way (check stabilization window in [Algorithm details][])
+
+[Algorithm details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 4: Scale Up As Usual, Do Not Scale Down
+
+This mode is essential when you don’t want to risk scaling down at all.
+
+Create an HPA with the following constraints:
+
+- `constraints`:
+  - `scaleDown`:
```

Comment:
This is a little clunky -- if there are ever new modes here (I don't have even a hypothetical example, but pretend :), then any YAML that carries this payload will not have that new field.

There are other API conventions being strengthened around one-of sets, which want a discriminator field, so maybe this looks like:

```
constraints:
  scaleDown:  # 3 pods per 10 minutes
    policy: Pods
    value: 3
    periodSeconds: 600
```

```
constraints:
  scaleDown:  # no scale-down allowed
    policy: Disabled
```

Or if pods and percent are not mutually exclusive (?) then something like:

```
constraints:
  scaleDown:  {} # no scale-down allowed
```

This changes the defaults from being on the leaf fields to being on the struct.  If `scaleDown` is not specified, the default value is `{ percent: 100, pods: 4 }`, but if it is specified, the fields inside default to 0.

Something like that.

Reply:
@gliush this is worth considering for https://github.com/kubernetes/kubernetes/pull/74525. Just wanted to make sure you didn't miss this feedback.